linkerd/distribute/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+        Self::Service {
+            inner: target.param().into(),
+            backends: self.backends.clone(),
+            ready: None,
+        }
+    }
+}
+
+// === impl Distribute ===
+
+impl<K: Clone, S: Clone> Clone for Distribute<K, S> {
+    fn clone(&self) -> Self {
+        Self {
+            backends: self.backends.clone(),
+            inner: self.inner.clone(),
+            // Clear the ready index so that the new clone must become ready
+            // independently.
+            ready: None,
+        }
+    }
+}
+
+impl<Req, K, S> Service<Req> for Distribute<K, S>
+where
+    K: Hash + Eq + std::fmt::Debug,
+    S: Service<Req> + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    // Note that this doesn't necessarily drive all inner services to readiness.
+    // We expect that these inner services should be buffered or otherwise drive
+    // themselves to readiness (i.e. via SpawnReady).
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // If we've already chosen a ready index, then skip polling.
+        if self.ready.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        match self.inner {
+            Inner::Empty => {
+                tracing::debug!("empty distribution will never become ready");
+            }
+
+            Inner::FirstAvailable { ref keys } => {
+                for key in keys.iter() {
+                    let svc = self
+                        .backends
+                        .get(key)
+                        .expect("distributions must not reference unknown backends");
+                    let mut svc = svc.clone();
+                    if svc.poll_ready(cx)?.is_ready() {
+                        self.ready = Some(svc);
+                        return Poll::Ready(Ok(()));
+                    }
+                }
+            }
+
+            Inner::RandomAvailable {
+                ref keys,
+                ref mut polled_idxs,
+                ref mut rng,
+            } => {
+                // Choose a random index (via the weighted distribution) to try
+                // to poll the backend. Continue selecting endpoints until we
+                // find one that is ready or we've tried all backends in the
+                // distribution.
+                polled_idxs.clear();
+                while polled_idxs.len() != keys.keys.len() {
+                    let (idx, _, svc) = self
+                        .backends
+                        .get_full(keys.next(rng))
+                        .expect("distributions must not reference unknown backends");
+                    if polled_idxs.insert(idx) {
+                        // The index was not already polled, so poll it.
+                        let mut svc = svc.clone();
+                        if svc.poll_ready(cx)?.is_ready() {
+                            self.ready = Some(svc);
```

Comment:
same comment as in `FirstAvailable` regarding services with per-clone readiness state.